### PR TITLE
boards: sensry: Add ganymed_bob board support

### LINKED ISSUE #77757: [RISCV] support for SY1xx series SOC + breakout board
**Is your enhancement proposal related to a problem? Please describe.**

we do have a new SOC SY1xx (privileged M+U) based on RISCV together with a break-out-board for testing purposes. 

**Describe the solution you'd like**

we did prepare a (draft) PR that includes SOC and drivers for sysclock, interrupt, uart to provide basic functionality. would like to learn about what else is needed to add the new SOC + BOARD

**Describe alternatives you've considered**

actually there is no alternative, ... ;)

**Additional context**

PR includes

* board description
* board dts + yaml + Kconfig
* drivers
   * serial
   * timer
* dts
   *  bindings
      * cpu 
      * irq
      * serial
      * timer
   * riscv
      *  cpu 
* soc
   * Kconfig.soc
   * Kconfig
   * common (.c, .S)
      *  basic implementation of soc specifics

### PR DESCRIPTION
**Pull request by tswaehn**
Add soc support for Sensry's RISCV32 based SY1xx.
Variants of the soc are GBM and GEN1.
Add board support for eval board ganymed_bob, which is a break-out-board for both soc variants.

fixes: #77757

### CODE REVIEW

File: boards/sensry/ganymed_bob/board.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2024 sensry.io
+# SPDX-License-Identifier: Apache-2.0
+
+board:
+  name: ganymed_bob
+  vendor: sensry
+  socs:
+    - name: sy120_gbm
+    - name: sy120_gen1
```
Comment (nordicjm) [inline boards/sensry/ganymed_bob/board.yml:9]: why isn't this a variant if all it does is add sensors?

File: boards/sensry/ganymed_bob/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+Ganymed BOB
+############
+
+Overview
+********
+
+.. note::
+
+   All software for the nGanymed BOB is experimental and hardware availability
+   is restricted to the participants in the limited sampling program.
```
Comment (nordicjm) [inline boards/sensry/ganymed_bob/doc/index.rst:12]: this looks like you've taken the document from a nordic board, unless this is true, it should be removed

File: boards/sensry/ganymed_bob/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+ * SY120-GEN1 - Generic Module with top level sensors (Bosch BME680 - SPI1 , Bosch BMA456 - SPI0, Bosch BMG250 - SPI2, STMicro MIS2DH - I2C0)
+* power section for on-board power generation (selectable)
+* 40-pin JTAG connector (compatible to Olimex ARM-JTAG-OCD-H)
+* USB over FTDI (connected to UART0)
+* Header for I/Os and additional configuration
+
+Supported Features
+==================
+
+The variant ``ganymed-bob/sy120-gbm/cpuapp`` board configuration supports the following
```
Comment (nordicjm) [inline boards/sensry/ganymed_bob/doc/index.rst:67]: *board target, the board target is also actually wrong

File: boards/sensry/ganymed_bob/doc/index.rst [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+| UART      | on-chip    | serial               |
++-----------+------------+----------------------+
+| TSN       | on-chip    | ethernet MAC         |
++-----------+------------+----------------------+
+| CAN       | on-chip    | CAN                  |
++-----------+------------+----------------------+
+
+Other hardware features have not been enabled yet for this board.
+
+The variant ``ganymed-bob/sy120-gen1/cpuapp`` board configuration supports additionally to the ``ganymed-bob/sy120-gbm/cpuapp`` the following
```
Comment (nordicjm) [inline boards/sensry/ganymed_bob/doc/index.rst:102]: as above

File: boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.dts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+/* Copyright (c) 2024 sensry.io */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/dts-v1/;
+
+#include <sensry/ganymed-sy1xx.dtsi>
+
+/ {
+
+
```
Comment (nordicjm) [inline boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.dts:10]: excess spacing

File: boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+identifier: ganymed_bob_sy120_gbm
+vendor: sensry
+name: Ganymed family breakout board
+type: mcu
+arch: riscv
+toolchain:
+  - zephyr
+
```
Comment (nordicjm) [inline boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.yaml:11]:
```suggestion
```

File: boards/sensry/ganymed_bob/ganymed_bob_sy120_gen1.dts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+/* Copyright (c) 2024 sensry.io */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/dts-v1/;
+
+#include <sensry/ganymed-sy1xx.dtsi>
+
+/ {
+
+
```
Comment (nordicjm) [inline boards/sensry/ganymed_bob/ganymed_bob_sy120_gen1.dts:10]: as above

File: boards/sensry/ganymed_bob/ganymed_bob_sy120_gen1.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+identifier: ganymed_bob_sy120_gen1
+vendor: sensry
+name: Ganymed family breakout board
+type: mcu
+arch: riscv
+toolchain:
+  - zephyr
+
```
Comment (nordicjm) [inline boards/sensry/ganymed_bob/ganymed_bob_sy120_gen1.yaml:11]: as above

File: drivers/serial/Kconfig.sy1xx [hunk 9]
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2024 sensry.io
+
+config UART_SY1XX
+	bool "Sensry SY1xx driver"
+	default y
+    depends on DT_HAS_SENSRY_SY1XX_UART_ENABLED
```
Comment (nordicjm) [inline drivers/serial/Kconfig.sy1xx:7]: tabs in Kconfig

File: soc/sensry/ganymed/Kconfig [hunk 10]
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2024 sensry.io
+
+config SOC_FAMILY_GANYMED
+	select BUILD_OUTPUT_HEX
+	select BUILD_OUTPUT_BIN
```
Comment (nordicjm) [inline soc/sensry/ganymed/Kconfig:6]: why are both of these always required?

File: soc/sensry/ganymed/Kconfig.soc [hunk 11]
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2024 sensry.io
+
+config SOC_SERIES_SY1XX
+    bool
+    select SOC_FAMILY_GANYMED
+
+config SOC_SERIES
+	default "sy1xx" if SOC_SERIES_SY1XX
+
```
Comment (nordicjm) [inline soc/sensry/ganymed/Kconfig.soc:10]: 1x empty line not 2

File: soc/sensry/ganymed/sy1xx/Kconfig [hunk 12]
```diff
@@ -0,0 +1,147 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2024 sensry.io
+
+config SOC_SERIES_SY1XX
+    bool
+	select RISCV
+    select RISCV_ISA_RV32I
```
Comment (nordicjm) [inline soc/sensry/ganymed/sy1xx/Kconfig:7]: wrong indent

File: soc/sensry/ganymed/sy1xx/Kconfig [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    select RISCV_ISA_EXT_M
+    select RISCV_ISA_EXT_ZICSR
+    select RISCV_PRIVILEGED
+
+
+if SOC_SERIES_SY1XX
+
+config RISCV_PMP
```
Comment (nordicjm) [inline soc/sensry/ganymed/sy1xx/Kconfig:15]: these go in the `Kconfig.defconfig` file

File: soc/sensry/ganymed/sy1xx/Kconfig [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+config TIMESLICE_PER_THREAD
+    default y
+
+config TIMESLICE_SIZE
+    default 10
+
+
+## general
+config BOOT_BANNER
+    default y
```
Comment (nordicjm) [inline soc/sensry/ganymed/sy1xx/Kconfig:95]: what?

File: soc/sensry/ganymed/sy1xx/Kconfig [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+config EARLY_CONSOLE
+    default y
+
+
+## shell
+config SHELL
+    default y
+config KERNEL_SHELL
+    default y
+config SHELL_BACKEND_SERIAL
```
Comment (nordicjm) [inline soc/sensry/ganymed/sy1xx/Kconfig:138]: No to any of this

File: soc/sensry/ganymed/sy1xx/common/event-unit.h [hunk 16]
```diff
@@ -0,0 +1,8 @@
+//
+// Created by tswaehn on 8/28/24.
+//
+
+#ifndef MY_ZEPHYR_APP_LOCAL_ZEPHYRPROJECT_LATEST_ZEPHYR_SOC_SENSRY_GANYMED_SY1XX_COMMON_EVENT_UNIT_H
+#define MY_ZEPHYR_APP_LOCAL_ZEPHYRPROJECT_LATEST_ZEPHYR_SOC_SENSRY_GANYMED_SY1XX_COMMON_EVENT_UNIT_H
+
```
Comment (nordicjm) [inline soc/sensry/ganymed/sy1xx/common/event-unit.h:7]: file can go

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: boards/sensry/ganymed_bob/board.yml (hunk 1 above)
Comment (tswaehn) [reply boards/sensry/ganymed_bob/board.yml:9]: as the sensors are placed inside the chip, we can use the chip on different boards. so the sensors come with the chip and not with the board.

File: boards/sensry/ganymed_bob/doc/index.rst (hunk 2 above)
Comment (tswaehn) [reply boards/sensry/ganymed_bob/doc/index.rst:12]: yes, its taken from nordic template. its by intention. we start to introduce the chip.

File: boards/sensry/ganymed_bob/board.yml (hunk 1 above)
Comment (nordicjm) [reply boards/sensry/ganymed_bob/board.yml:9]: 2 SoCs it is then

File: soc/sensry/ganymed/sy1xx/Kconfig (hunk 12 above)
Comment (tswaehn) [reply soc/sensry/ganymed/sy1xx/Kconfig:7]: fixed

File: soc/sensry/ganymed/sy1xx/Kconfig (hunk 14 above)
Comment (tswaehn) [reply soc/sensry/ganymed/sy1xx/Kconfig]: removed

File: soc/sensry/ganymed/sy1xx/Kconfig (hunk 15 above)
Comment (tswaehn) [reply soc/sensry/ganymed/sy1xx/Kconfig]: moved this to the board, should not be in soc

File: soc/sensry/ganymed/sy1xx/common/event-unit.h (hunk 16 above)
Comment (tswaehn) [reply soc/sensry/ganymed/sy1xx/common/event-unit.h]: removed

File: boards/sensry/ganymed_bob/Kconfig.defconfig [hunk 17]
```diff
@@ -0,0 +1,2 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2024 sensry.io
```
[the remark this answers is not in the document]
Comment (tswaehn) [reply boards/sensry/ganymed_bob/Kconfig.defconfig:2]: done

File: boards/sensry/ganymed_bob/doc/index.rst (hunk 3 above)
Comment (tswaehn) [reply boards/sensry/ganymed_bob/doc/index.rst:67]: fixed

File: boards/sensry/ganymed_bob/doc/index.rst (hunk 4 above)
Comment (tswaehn) [reply boards/sensry/ganymed_bob/doc/index.rst:102]: fixed

File: boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.dts (hunk 5 above)
Comment (tswaehn) [reply boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.dts:10]: fixed

File: boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.yaml (hunk 6 above)
Comment (tswaehn) [reply boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.yaml:11]: fixed

File: boards/sensry/ganymed_bob/ganymed_bob_sy120_gen1.dts (hunk 7 above)
Comment (tswaehn) [reply boards/sensry/ganymed_bob/ganymed_bob_sy120_gen1.dts:10]: fixed

File: drivers/serial/Kconfig.sy1xx (hunk 9 above)
Comment (tswaehn) [reply drivers/serial/Kconfig.sy1xx:7]: fixed

File: soc/sensry/ganymed/Kconfig (hunk 10 above)
Comment (tswaehn) [reply soc/sensry/ganymed/Kconfig]: actually really need only binary. removed the hex options

File: soc/sensry/ganymed/Kconfig.soc (hunk 11 above)
Comment (tswaehn) [reply soc/sensry/ganymed/Kconfig.soc:10]: fixed

File: soc/sensry/ganymed/sy1xx/Kconfig (hunk 13 above)
Comment (tswaehn) [reply soc/sensry/ganymed/sy1xx/Kconfig:15]: done

**Comment by tswaehn** [comment]
@nordicjm I think formal, code style related things should be covered now. its actually great you spent the time for a feedback. thank you!

File: boards/sensry/ganymed_bob/Kconfig.defconfig [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+config TIMESLICING
+	default y
+config TIMESLICE_PER_THREAD
+	default y
+config TIMESLICE_SIZE
+	default 10
+
+## serial
+config SERIAL_SUPPORT_INTERRUPT
+	default n
```
Comment (dcpleung) [inline boards/sensry/ganymed_bob/Kconfig.defconfig:18]: This should be "selected" by driver and no need to explicitly disable it.

File: boards/sensry/ganymed_bob/Kconfig.defconfig [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+config UART_ASYNC_API
+	default n
```
Comment (dcpleung) [inline boards/sensry/ganymed_bob/Kconfig.defconfig:20]: `CONFIG_UART_ASYNC_API` is disabled by default so to explicitly do so.

File: boards/sensry/ganymed_bob/Kconfig.defconfig [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+config SERIAL_HAS_DRIVER
+	default y
```
Comment (dcpleung) [inline boards/sensry/ganymed_bob/Kconfig.defconfig:22]: This selected by driver and no need to explicitly enable it.

File: drivers/serial/uart_sy1xx.c [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+
+	drivers_uart_read(dev, &dummy_request);
+	return 0;
+}
+
+/*
+ * \return: < 0 ... error
+ *	    = 0 ok
+ *	    > 0 ... busy
+ */
```
Comment (dcpleung) [inline drivers/serial/uart_sy1xx.c:106]: I recommend changing this into proper doxygen style comments so IDE can display it during code completion, especially the return values.

ditto for those below.

**Review by dcpleung — COMMENTED** [review]
Could you split the commits into small ones? For example, one for timer driver, one for serial driver, one for SoC, one for board and one for dts? That would make it easier to review.

File: drivers/serial/uart_sy1xx.c (hunk 21 above)
Comment (tswaehn) [reply drivers/serial/uart_sy1xx.c]: done

File: boards/sensry/ganymed_bob/Kconfig.defconfig (hunk 20 above)
Comment (tswaehn) [reply boards/sensry/ganymed_bob/Kconfig.defconfig]: understand

File: boards/sensry/ganymed_bob/Kconfig.defconfig (hunk 19 above)
Comment (tswaehn) [reply boards/sensry/ganymed_bob/Kconfig.defconfig]: fixed

**Comment by tswaehn** [comment]
dear @nordicjm, dear @dcpleung 

I followed you advice. Would be great to have your review of my changes.

File: dts/bindings/cpu/sensry,ganymed-sy1xx.yaml [hunk 22]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2024 sensry.io
+# SPDX-License-Identifier: Apache-2.0
+
+description: Sensry Ganymed SY1xx Core CPU
+
+compatible: "sensry,sy1xx"
```
Comment (nordicjm) [inline dts/bindings/cpu/sensry,ganymed-sy1xx.yaml:6]: missing cpu.yaml include, probably missing riscv,cpus.yaml include too

File: soc/sensry/ganymed/sy1xx/Kconfig [hunk 23]
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2024 sensry.io
+
+config SOC_SERIES_SY1XX
+	bool
```
Comment (nordicjm) [inline soc/sensry/ganymed/sy1xx/Kconfig:5]:
```suggestion
```

File: boards/sensry/ganymed_bob/Kconfig.defconfig (hunk 17 above)
[the remark this answers is not in the document]
Comment (nordicjm) [reply boards/sensry/ganymed_bob/Kconfig.defconfig:2]: ??

File: boards/sensry/ganymed_bob/doc/img/ganymed_bob_sy120_gbm.webp
Comment (nordicjm) [inline boards/sensry/ganymed_bob/doc/img/ganymed_bob_sy120_gbm.webp:1]: Put webp files through https://tinypng.com/ to reduce file size

File: boards/sensry/ganymed_bob/Kconfig.defconfig (hunk 17 above)
[the remark this answers is not in the document]
Comment (tswaehn) [reply boards/sensry/ganymed_bob/Kconfig.defconfig]: missed that again. sure, can go.

File: boards/sensry/ganymed_bob/doc/img/ganymed_bob_sy120_gbm.webp
Comment (tswaehn) [reply boards/sensry/ganymed_bob/doc/img/ganymed_bob_sy120_gbm.webp:1]: actually I am impressed. I was using GIMP before - which is THE image tool. and finally tinypng shrinked them by 45% without visual impact.

File: dts/bindings/cpu/sensry,ganymed-sy1xx.yaml (hunk 22 above)
Comment (tswaehn) [reply dts/bindings/cpu/sensry,ganymed-sy1xx.yaml:8]: riscv,cpus.yaml includes cpu.yaml; included riscv,cpus.yaml

**Comment by tswaehn** [comment]
I am really keen in getting the PR accepted. Just wondering if there is things open that I need to care about?

**Comment by tswaehn** [comment]
I think to have covered all points from review. Is there anything I could do?

**Review by nordicjm — APPROVED** [review]
Board/SoC files OK

**Review by dcpleung — APPROVED** [review]

File: soc/sensry/ganymed/sy1xx/CMakeLists.txt [hunk 24]
```diff
@@ -0,0 +1,16 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2024 sensry.io
+
+zephyr_include_directories(common)
+
+zephyr_compile_options(-march=rv32imc_zicsr -mabi=ilp32)
```
Comment (dcpleung) [inline soc/sensry/ganymed/sy1xx/CMakeLists.txt:6]: `cmake/compiler/gcc/target_riscv.cmake` would generate the correct `march` argument if you set `CONFIG_RISCV_ISA_EXT_*` correctly. So there is no need to specify `march` here.

**Comment by dcpleung** [comment]
Another suggestion is to prefix the SoC macro like `UDMA_*` and `PAD_CONFIG` with the SoC name (or other more unique identifier). This would help others when searching for code as they can exclude such substrings.

**Comment by fkokosinski** [comment]
Question: what about the maintainership status of this board? If I understand correctly, right now it's not covered by any of the preexisting areas of maintenance.

**Comment by tswaehn** [comment]
> Question: what about the maintainership status of this board? If I understand correctly, right now it's not covered by any of the preexisting areas of maintenance.

I can do, ... just need to understand my duties.

File: soc/sensry/ganymed/sy1xx/CMakeLists.txt (hunk 24 above)
Comment (tswaehn) [reply soc/sensry/ganymed/sy1xx/CMakeLists.txt:6]: @dcpleung just added PR here: https://github.com/zephyrproject-rtos/zephyr/pull/78633

Comment (tswaehn) [reply soc/sensry/ganymed/sy1xx/CMakeLists.txt:6]: this is merged, would it be ok to "Resolve Conversation" ?

**Pull request merged**